### PR TITLE
[SPARK-42552][SQL] Fix select without parentheses can't be parsed

### PR DESCRIPTION
**Pull request by Hisoka-X**
### What changes were proposed in this pull request?
The Sql Parser will try use `SSL`(faster) mode parse sqlText at first, if throw `ParseCancellationException` then try parse sqlText with `LL`(more correcter) mode. But Spark use custom `SparkParserErrorStrategy` will not throw `ParseCancellationException` but `ParseException` in some situation. So we should catch `ParseException` then to make sure parser use `LL` mode try again.

Fix sql parser throw `ParseException` in `SSL` mode will not try `LL` mode.

### Why are the changes needed?
Make some special sql can be parsed. Like `SELECT 1 UNION SELECT 1`.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Add new test

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ParseDriver.scala [hunk 1]
```diff
@@ -127,7 +127,7 @@ abstract class AbstractSqlParser extends ParserInterface with SQLConfHelper with
         toResult(parser)
       }
       catch {
-        case e: ParseCancellationException =>
+        case _: ParseCancellationException | _: ParseException =>
```
Comment (pan3793) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ParseDriver.scala:130]: This fix looks too broad.

Have done some investigations, the key point should be to use `BailErrorStrategy`.

This should be a long-standing issue, before SPARK-38385, Spark uses `DefaultErrorStrategy`, and after SPARK-38385 Spark uses `class SparkParserErrorStrategy() extends DefaultErrorStrategy`.

So this is not the correct implementation of the "two-stage parsing strategy" 

https://github.com/antlr/antlr4/issues/192#issuecomment-15238595

**Comment by LuciferYang** [comment]
> Make some special sql can be parsed. Like SELECT 1 UNION SELECT 1.

cc @wangyum FYI

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala [hunk 2]
```diff
@@ -197,6 +197,12 @@ class PlanParserSuite extends AnalysisTest {
       parameters = Map.empty)
   }
 
+  test("select and union without parentheses") {
+    val plan = Distinct(OneRowRelation().select(Literal(1))
+      .union(OneRowRelation().select(Literal(1))))
+    assertEqual("select 1 union select 1", plan)
```
Comment (pan3793) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala:203]: Have done a quick check w/ the following change

```patch
- parser.setErrorHandler(new SparkParserErrorStrategy())
+ parser.setErrorHandler(new BailErrorStrategy())
```

```
sbt:spark-catalyst> testOnly *PlanParserSuite -- -z "select and union without parentheses"
[warn] multiple main classes detected: run 'show discoveredMainClasses' to see the list
[info] compiling 1 Scala source to /Users/chengpan/Projects/apache-spark/sql/catalyst/target/scala-2.12/classes ...
[info] PlanParserSuite:
[info] - select and union without parentheses (364 milliseconds)
[info] Run completed in 1 second, 463 milliseconds.
[info] Total number of tests run: 1
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 1, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
[success] Total time: 31 s, completed Apr 17, 2023 10:05:54 PM
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ParseDriver.scala (hunk 1 above)
Comment (Hisoka-X) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ParseDriver.scala:130]: Thanks for remind. The `SparkParserErrorStrategy` will rewrap Exception which throw by antlr4. And finally throw it by `ParseException`. So if we want to fix this, catch `ParserException` will be a way. If you have any better idea please let me know. Thanks!

**Comment by HyukjinKwon** [comment]
cc @MaxGekk and @gengliangwang FYI

**Comment by cloud-fan** [comment]
Can you provide more background to help people understand and review? I have no idea why catching one more exception type can fix this parser issue.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ParseDriver.scala (hunk 1 above)
Comment (pan3793) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ParseDriver.scala:130]: @Hisoka-X @cloud-fan I think we should provide two `ErrorStrategy`s
```
class SparkParserBailErrorStrategy() extends BailErrorStrategy
class SparkParserDefaultErrorStrategy() extends DefaultErrorStrategy
```

and following the https://github.com/antlr/antlr4/issues/192#issuecomment-15238595 to implement the "two-stage parsing strategy"

**Comment by Hisoka-X** [comment]
> Can you provide more background to help people understand and review? I have no idea why catching one more exception type can fix this parser issue.

Thanks for point that. I updated the description.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ParseDriver.scala (hunk 1 above)
Comment (Hisoka-X) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ParseDriver.scala:130]: > @Hisoka-X @cloud-fan I think we should provide two `ErrorStrategy`s
> 
> ```
> class SparkParserBailErrorStrategy() extends BailErrorStrategy
> class SparkParserDefaultErrorStrategy() extends DefaultErrorStrategy
> ```
> 
> and follow the [antlr/antlr4#192 (comment)](https://github.com/antlr/antlr4/issues/192#issuecomment-15238595) to implement the "two-stage parsing strategy"

I'm not sure this way will be worked. The main reason for custom `ErrorStrategy` are replace anltr4 Exception by some Spark Exception. So we need catch the Spark Exception too.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ParseDriver.scala:130]: @pan3793 can you make a PR to implement your idea and see if it works?

**Comment by Hisoka-X** [comment]
Replace by https://github.com/apache/spark/pull/40835

**Pull request closed without merging**